### PR TITLE
EVA-2565 - Handle merges and splits during remapping

### PR DESCRIPTION
**Pull request by sundarvenkata-EBI**
Proposed strategy [here](https://docs.google.com/spreadsheets/d/1KQLVCUy-vqXKgkCDt2czX6kuMfsjfCc9uBsS19MZ6dY/edit#rangeid=623655670)

### CODE REVIEW

**Review by tcezard — APPROVED** [review]
This looks great.

**Review by nitin-ebi — APPROVED** [review]

File: eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/ClusteringWriter.java [hunk 1]
```diff
@@ -140,7 +163,7 @@ public void write(List<? extends SubmittedVariantEntity> submittedVariantEntitie
     }
 
     private void getOrCreateClusteredVariantAccessions(List<? extends SubmittedVariantEntity> submittedVariantEntities)
-            throws AccessionCouldNotBeGeneratedException {
+            throws AccessionCouldNotBeGeneratedException, AccessionDoesNotExistException {
         if (processClusteredRemappedVariants) {
             Set<SubmittedVariantEntity> processedSubmittedVariants = processClusteredRemappedVariants(submittedVariantEntities);
             submittedVariantEntities = Collections.unmodifiableList(submittedVariantEntities.stream()
```
[the remark this answers is not in the document]
Comment (sundarvenkata-EBI) [reply eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/ClusteringWriter.java]: Addressed

File: eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/RSSplitWriter.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+                                        ClusteredVariantEntity clusteredVariantEntity) {
+        // Choose which operation collection to write to: EVA or dbSNP
+        Class<? extends EventDocument<IClusteredVariant, Long, ? extends ClusteredVariantInactiveEntity>>
+                operationClass = clusteringWriter.getClusteredOperationCollection(oldRSAccession);
+        ClusteredVariantOperationEntity splitOperation = new ClusteredVariantOperationEntity();
+        String splitOperationDescription = "Due to hash mismatch, rs" + newRSAccession +
+                " was issued to split from rs" + oldRSAccession + ".";
+        splitOperation.fill(EventType.RS_SPLIT, oldRSAccession, newRSAccession, splitOperationDescription,
+                            Collections.singletonList(new ClusteredVariantInactiveEntity(clusteredVariantEntity)));
+        this.mongoTemplate.insert(splitOperation, this.mongoTemplate.getCollectionName(operationClass));
```
[the remark this answers is not in the document]
Comment (sundarvenkata-EBI) [reply eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/RSSplitWriter.java]: Addressed

File: eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/ClusteringWriter.java [hunk 3]
```diff
@@ -447,22 +471,34 @@ private void checkForMerges(List<? extends SubmittedVariantEntity> submittedVari
         }
     }
 
-    private void merge(Long providedAccession, String hash, Long accessionInDatabase) {
+    protected void merge(Long providedAccession, String hash, Long accessionInDatabase)
```
[the remark this answers is not in the document]
Comment (sundarvenkata-EBI) [reply eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/ClusteringWriter.java:478]: At this time, I would like to hold-off on large-scale refactoring. I can address it in EVA-2608.

**Review by tcezard — APPROVED** [review]
Thank you for adding the counts

**Pull request merged**